### PR TITLE
Convert website dark mode to presets

### LINKED ISSUE #645: Add the ability to customize all parts of the Scratch Website
**Is your feature request related to a problem? Please describe.**

Add an option to change what the actual Scratch website looks like, as in colours and fonts (add the ability to upload fonts from the user's PC).

**Describe the solution you'd like**

The option to change colours and fonts of any part of the website.

**Describe alternatives you've considered**

null

**Additional context**

null

### LINKED ISSUE #1833: Website dark mode doesn't make scrollbars dark in forum preview
**Describe the bug**

Website dark mode doesn't inject `scrollbar.css` into the forum preview, so scrollbars aren't dark there.

**To Reproduce**

1. Go to a forum topic
2. Copy and paste `[code]long text long text long text long text long text long text long text long text long text long text long text long text long text long text long text long text[/code]` into the reply box
3. Click the preview button (green ✅)
4. The scrollbar in the preview box isn't dark.

**Expected behavior**

The scrollbar would be dark.

**Screenshots**

A screenshot of the bug

### LINKED ISSUE #2740: Dark mode creates extra margin between projects in mystuff
https://user-images.githubusercontent.com/72760579/121617585-4c77e200-ca33-11eb-9835-2cbfd067a3bb.mp4

### LINKED ISSUE #2741: Dark mode makes eye icon weird
https://user-images.githubusercontent.com/72760579/121617987-02433080-ca34-11eb-8779-b14e16adc86c.mp4

### LINKED ISSUE #3079: Custom/Preset Page Colors
### Is your request related to a problem? Please describe.
With the new studio update, studios have a blue background, which has shown to be hard to look at for users (such as myself). To me, that's the biggest visual issue with the update.
Of course, with Scratch Addons, you can switch on Dark Mode or Old Studio Layout, but I prefer the 3.0 layout over the addon style, as well as the light mode over the dark mode.

### Describe the solution you'd like
I would love to have the ability to set the page colors to a custom setting, or at least have the ability to set the background color of studio pages to the white seen across Scratch. I greatly enjoy the ability to change stuff like the header color, and I think being able to edit the background would be useful as well, even if it's just presets or maybe even an just an option in the Old Studio Layout addon to make the background white.

### LINKED ISSUE #3829: Dark mode is broken when becoming scratcher
### Description
When i was becoming a scratcher on my other account that Im not telling you about i saw that the css was broken

### Repoduction steps

1.  create an account
2.  turn on the dark mode addon
3.  become a scratcher

### Expected behaviour

The text boxes should be filled with white and the boxes at the bottom should be white too
### Actual behaviour

the textboxes and boxes aren't white

### Screenshots
image
image
image
image
image
image

### Environment

- Browser:  Firefox 94
- Scratch Addons version:  1.22.1
- Operating system:  Windows 10

### PR DESCRIPTION
**Pull request by mxmou**
Resolves #645 (not the fonts part, but there is another issue for that: #1646)
Adds a separator between comments as suggested in #792
Resolves #1833
Resolves #2740
Resolves #2741
Resolves #3079
Resolves #3829

### Dark mode changes

Adds color settings to website dark mode, with presets for the current "Experimental Dark" style and Scratch defaults. Dark WWW (#517) is also a preset.
image

On scratchr2 pages, variants of the selected colors are generated to make sure that the Scratch's default colors preset doesn't change those pages.

To do:
* Find and fix bugs
* [x] Make the colored areas setting affect banners added by Scratch 2.0 → 3.0
* Migrate existing settings (#3861)

### Other changes

* The `threshold` property of a `textColor` custom CSS variable can be a color provider
* `brighten` properly supports changing the opacity of a color
* Fixed the `isNotScratchWWW` regular expression not matching `/educators/register` and `/educators/waiting`
* Slightly changed the design of `message-filters` and `project-notes-tabs`

### CODE REVIEW

**Comment by WorldLanguages** [comment]
Maybe it's just me, and users don't really care that much, but could there be a preset that looks like classic dark-www (assuming it's possible now that every color is a setting, instead of the CSS selectors being hardcoded)? It will take me a bit to get used to comment sections having a background color that's almost black, even if it's consistent with the other Experimental Dark preset. We could switch users to the new one and make them get used to it, but if possible we should let users go back in the settings page if they really want to.

Aiming to release this for v1.22.0

**Comment by WorldLanguages** [comment]
By the way mxmou, please confirm whether you fixed the scratcher promotion unreadable text bug on this PR

**Comment by RedGuy12** [comment]
Was transition code ever added for the old scratchr2 settings?


________________________________

**Comment by mxmou** [comment]
> Maybe it's just me, and users don't really care that much, but could there be a preset that looks like classic dark-www (assuming it's possible now that every color is a setting, instead of the CSS selectors being hardcoded)? It will take me a bit to get used to comment sections having a background color that's almost black, even if it's consistent with the other Experimental Dark preset. We could switch users to the new one and make them get used to it, but if possible we should let users go back in the settings page if they really want to.

I think it would be better to choose a color that's more similar to the old one instead of having two very similar presets.

> By the way mxmou, please confirm whether you fixed the scratcher promotion unreadable text bug on this PR

Not yet.

**Comment by WorldLanguages** [comment]
> Was transition code ever added for the old scratchr2 settings?

It's a pending TODO on this PR.

**Comment by WorldLanguages** [comment]
Why the milestone change? @TheColaber

**Comment by WorldLanguages** [comment]
> I think it would be better to choose a color that's more similar to the old one instead of having two very similar presets.

Sorry I think I misunderstood this when I first read it - do you mean we should change the default color, or that users should choose a similar color to the old one if they dislike the update?

**Comment by WorldLanguages** [comment]
Promotion is now readable:
image

If we really want to, we could change the background-image of the after pseudoelement with one that has transparent background (URL is https://cdn.scratch.mit.edu/scratchr2/static/__fd54d8440ae146e24183f3a920f85c36__/images/bubble-bottom-southwest.png)

**Comment by mxmou** [comment]
> > I think it would be better to choose a color that's more similar to the old one instead of having two very similar presets.
> 
> Sorry I think I misunderstood this when I first read it - do you mean we should change the default color, or that users should choose a similar color to the old one if they dislike the update?

I meant that we should choose a different color. Keeping the old one and changing the editor Experimental Dark to match might be a better idea than what I initially did on this PR.

> Promotion is now readable: (image)

That's great. Does changing making the content background or gray background white without changing any other settings break anything?

**Comment by WorldLanguages** [comment]
> I meant that we should choose a different color. Keeping the old one and changing the editor Experimental Dark to match might be a better idea than what I initially did on this PR.

Honestly, I don't think the colors need to strictly match either.

> That's great. Does changing making the content background or gray background white without changing any other settings break anything?

Can you clarify?

**Comment by WorldLanguages** [comment]
Maybe we should move this to v1.23.0 milestone in order to have more time to discuss the colors?

**Comment by mxmou** [comment]
> > That's great. Does changing making the content background or gray background white without changing any other settings break anything?
> 
> Can you clarify?

I'm asking if text is still readable when setting content background to white and gray background to the default color, or gray background to white and content background to the default.

**Comment by mxmou** [comment]
Would these colors in editor dark mode look good?
image

The current ones look like this:
image

**Comment by RedGuy12** [comment]
Yes I would use that


________________________________

**Comment by WorldLanguages** [comment]
They look good to me.
But do the colors really need to match between the editor "Experimental dark" theme and the website's? Does scratch-www use the same color palette from the Scratch editor? I don't think so (correct me if I'm wrong). Then why would we need to be consistent on something Scratch isn't consistent in?

**Comment by mxmou** [comment]
> They look good to me. But do the colors really need to match between the editor "Experimental dark" theme and the website's? Does scratch-www use the same color palette from the Scratch editor? I don't think so (correct me if I'm wrong). Then why would we need to be consistent on something Scratch isn't consistent in?

The two themes named "Experimental Dark" are meant to use similar colors (scratch-gui and scratch-www color palettes aren't exactly the same, but they're both based on the same main colors, so they're quite similar). The dark blue background in the editor doesn't look very good, so I think changing it to match the website theme would be an improvement. The settings would need to be updated for users who already use Experimental Dark.

**Comment by WorldLanguages** [comment]
@mxmou Okay let's do that then. After it's done we can discuss whether we want an `"Experimental" Dark (legacy)` preset for the editor or not.

**Comment by WorldLanguages** [comment]
@mxmou We moved migration to a different issue, so  this can be merged if the conflict is resolved.

**Comment by WorldLanguages** [comment]
This addon will be under "new and updated addons" after #3811 is merged.

**Comment by WorldLanguages** [comment]
The schema needs to be updated to reflect the new possibilities on `customCssVariables`.

File: addons/dark-www/scrollbar.css [hunk 1]
```diff
@@ -1,20 +1,20 @@
 body:not(.sa-body-editor)::-webkit-scrollbar,
 body:not(.sa-body-editor) ::-webkit-scrollbar {
-  width: 16px;
-  height: 16px;
+  width: 17px;
+  height: 17px;
```
Comment (WorldLanguages) [inline addons/dark-www/scrollbar.css:4]: @mxmou Just wondering why? 😂

Comment (mxmou) [reply addons/dark-www/scrollbar.css:4]: 17px is the width of Chromium's default scrollbar for some reason, so using the same width prevents the entire page from shifting slightly when enabling or disabling the setting.

File: addons/dark-www/experimental_scratchr2.css [hunk 2]
```diff
@@ -1,170 +1,660 @@
...
+  background-color: transparent !important;
+}
+.dropdown .dropdown-menu {
   background-clip: border-box;
+  border-color: var(--darkWww-border-20);
+  border-top-color: var(--darkWww-border-5);
+}
+.djangobb hr {
+  background-color: var(--darkWww-border-20);
+}
```
Comment (WorldLanguages) [inline addons/dark-www/experimental_scratchr2.css:545]: @mxmou
From feedback:
> The signature line uses the border colour, I have borders hidden to make the UI cleaner
i wanna remove the other borders but not that one
i still wanna hide the borders for a cleaner ui while showing the signature line for clarity

Comment (WorldLanguages) [reply addons/dark-www/experimental_scratchr2.css:545]: Personally I believe the request is reasonable, I wouldn't consider the signature line to be a "border". And I believe making borders 100% transparent might be pretty common.
But I don't have any concrete suggestions on how else we could compute an appropriate color for the signature line. Of course, we wouldn't want to add a specific setting for it.

Comment (mxmou) [reply addons/dark-www/experimental_scratchr2.css:545]: One option is to generate a different color for the signature line if the border color is fully transparent.

Comment (WorldLanguages) [reply addons/dark-www/experimental_scratchr2.css:545]: @mxmou Sounds good to me. Should I create an issue?

Comment (Secret-chest) [reply addons/dark-www/experimental_scratchr2.css:545]: What about making it 50% opaque text colour?

Comment (WorldLanguages) [reply addons/dark-www/experimental_scratchr2.css:545]: @Secret-chest Keep in mind that the color would need to stay the same under "Scratch's default colors" preset

**Pull request merged**